core/drivers/stm32_bsec.c

```diff
@@ -0,0 +1,510 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2017-2018, STMicroelectronics
+ */
+
+#include <assert.h>
+#include <drivers/stm32_bsec.h>
+#include <io.h>
+#include <kernel/generic_boot.h>
+#include <kernel/spinlock.h>
+#include <limits.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <stdint.h>
+#include <stm32_util.h>
+#include <util.h>
+
+struct bsec_dev {
+	struct io_pa_va base;
+	unsigned int upper_base;
+	unsigned int max_id;
+	bool closed_device;
+};
+
+/* Only 1 instance of BSEC is expected per platform */
+static struct bsec_dev bsec_dev;
+
+/* BSEC access protection */
+static unsigned int lock = SPINLOCK_UNLOCK;
+
+static uint32_t bsec_lock(void)
+{
+	return may_spin_lock(&lock);
+}
+
+static void bsec_unlock(uint32_t exceptions)
+{
+	may_spin_unlock(&lock, exceptions);
+}
+
+static uint32_t otp_max_id(void)
+{
+	return bsec_dev.max_id;
+}
+
+static uint32_t otp_bank_offset(uint32_t otp_id)
+{
+	assert(otp_id <= otp_max_id());
+
+	return ((otp_id & ~BSEC_OTP_MASK) >> BSEC_OTP_BANK_SHIFT) *
+		sizeof(uint32_t);
+}
+
+static uintptr_t bsec_base(void)
+{
+	return io_pa_or_va(&bsec_dev.base);
+}
+
+static uint32_t bsec_get_status(void)
+{
+	return read32(bsec_base() + BSEC_OTP_STATUS_OFF);
+}
+
+static uint32_t bsec_check_error(uint32_t otp_id)
+{
+	uint32_t bit = BIT(otp_id & BSEC_OTP_MASK);
+	uint32_t bank = otp_bank_offset(otp_id);
+
+	if (read32(bsec_base() + BSEC_DISTURBED_OFF + bank) & bit)
+		return BSEC_DISTURBED;
+
+	if (read32(bsec_base() + BSEC_ERROR_OFF + bank) & bit)
+		return BSEC_ERROR;
+
+	return BSEC_OK;
+}
+
+static uint32_t enable_power(void)
+{
+	size_t cntdown = 0;
+
+	io_mask32(bsec_base() + BSEC_OTP_CONF_OFF, BSEC_CONF_POWER_UP_MASK,
+		  BSEC_CONF_POWER_UP_MASK);
+
+	for (cntdown = BSEC_TIMEOUT_VALUE; cntdown; cntdown--)
+		if (bsec_get_status() & BSEC_MODE_PWR_MASK)
+			break;
+
+	return cntdown ? BSEC_OK : BSEC_TIMEOUT;
+}
+
+static uint32_t disable_power(void)
+{
+	size_t cntdown = 0;
+
+	io_mask32(bsec_base() + BSEC_OTP_CONF_OFF, 0, BSEC_CONF_POWER_UP_MASK);
+
+	for (cntdown = BSEC_TIMEOUT_VALUE; cntdown; cntdown--)
+		if (!(bsec_get_status() & BSEC_MODE_PWR_MASK))
+			break;
+
+	return cntdown ? BSEC_OK : BSEC_TIMEOUT;
+}
+
+static uint32_t bsec_power_safmem(bool enable)
+{
+	static unsigned int refcnt = UINT_MAX;
+	uint32_t result = BSEC_OK;
+	uint32_t exceptions = 0;
+
+	exceptions = bsec_lock();
+
+	/* Sync reference counter with effective state */
+	if (refcnt == UINT_MAX)
+		refcnt = bsec_get_status() & BSEC_MODE_PWR_MASK ?
```

Comment:
This is one of the situations where the ternary expression doesn't make it easier to read the code.

The logic seems to have changed since time I reviewed. Was that on purpose? Perhaps a comment to explain what should be the effective state would help?

Reply:
You're right, I replaced the initial `1` that is `SHREFCNT_NONSECURE_FLAG` with `SHREFCNT_SECURE_STEP` which is not what's expected. I'll add a comment to explain.

Reply:
Actually, the refcount can be simply initiated to 0 and BSEC treat as if default power-off. Fixed in a fixup commit.